usb: stm32: added usb_dc_detach code

StefJar opened the pull request:
Now detaching means, that the transfer endpoints are closed.
The usb device is stoped and deinit.

The usb periphal wil go to deepsleep after. That will save
some power.

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

Commit:
usb: stm32: added usb_dc_detach code

Now detaching means, that the transfer endpoints are closed.
The usb device is stoped and deinit.

The usb periphal wil go to deepsleep after. That will save
some power.

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

StefJar commented:
pls review
@erwango

loicpoulain commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -846,8 +846,13 @@ int usb_dc_ep_mps(const u8_t ep)
+	HAL_PCD_Stop(&usb_dc_stm32_state.pcd);
+	// take care that the endpoints are closed (usb_dc_ep_disable)
+	HAL_PCD_EP_Close(&usb_dc_stm32_state.pcd, EP0_IN);
+	HAL_PCD_EP_Close(&usb_dc_stm32_state.pcd, EP0_OUT);
```
is it strictly necessary? what about the other endpoints?

StefJar replied:
As far as I have seen there are only EP0_IN and EP0_OUT in that code. Guess stopping the usb, closing the endpoints and deinit is the most explicit and clean ending.

loicpoulain commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -846,8 +846,13 @@ int usb_dc_ep_mps(const u8_t ep)
+	HAL_PCD_EP_Close(&usb_dc_stm32_state.pcd, EP0_OUT);
+	// now we deinit
+	HAL_PCD_DeInit(&usb_dc_stm32_state.pcd);
 	return 0;
```
maybe we should create a usb_dc_stm32_deinit that undo everything done in usb_dc_stm32_init. and call it form here instead.

StefJar replied:
I am not sure about that. I am not a expert on the STM HAL and its different implementations. So you guys know best and should decide.

ydamigos reviewed (commented):
@StefJar How did you test the change? Which sample?
In PR #16026, when HAL_PCD_DeInit() was called, the application got stuck. Did you face a similar problem?
@jli157 FYI.

Commit:
usb: stm32: removed C++ style comments

removed C++ comments.

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

ydamigos commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -846,12 +846,9 @@ int usb_dc_ep_mps(const u8_t ep)
 int usb_dc_detach(void)
 {
-	// stop usb dev
 	HAL_PCD_Stop(&usb_dc_stm32_state.pcd);
```
We don't need do this as `HAL_PCD_DeInit()` calls `HAL_PDC_Stop()`

StefJar replied:
so let's break it down to only the `HAL_PCD_DeInit` call?
What is with the endpoints?

ydamigos replied:
> What is with the endpoints?

Probably application needs to close all the endpoints that it opened. As far as I understand, it depends when `usb_disable()` is called. 

I am searching out STM32 documentation for the correct sequence to disable USB device. So far no luck. @erwango Could you help on this?

StefJar replied:
> @StefJar How did you test the change? Which sample?
> In PR #16026, when HAL_PCD_DeInit() was called, the application got stuck. Did you face a similar problem?
> @jli157 FYI.

nope
it worked for me fine(on a stm32f412)

ydamigos replied:
> it worked for me fine(on a stm32f412)

How did you test it? Which sample or test case?

StefJar replied:
I have a sequence to put the device into sleep mode/reset. Inside that sequence I am deinit the usb. This works like follows:
- call `usb_disable`
- unint pins via `stm32_setup_pins`
- unclock usb via `LL_APB2_GRP1_DisableClock(LL_AHB2_GRP1_PERIPH_OTGFS)`

as you can see it is more less handcrafted. The test was run on 5 units (stm32f412) and went well

Commit:
usb: stm32: removed HAL_PCD_Stop

HAL_PCD_Stop is called inside HAL_PCD_Deinit so we can remove it. Moved endpoint close after usb device deinit.

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

StefJar commented:
will put everything into one commit and resubmit it

StefJar commented:
the current pull can be found here
https://github.com/zephyrproject-rtos/zephyr/pull/22767
